File: sdk/js/mirrordnaClient.js
/**
 * MirrorDNA Developer SDK - JavaScript Client
 *
 * Simple, local-only client for MirrorDNA protocol operations.
 * Provides easy-to-use interface for vault config loading, state hashing, and timeline validation.
 *
 * This SDK is designed for:
 * - Local development and testing
 * - Understanding MirrorDNA concepts
 * - Building simple Node.js integrations
 *
 * For production use, see the full TypeScript SDK in ../javascript/
 */

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

/**
 * Simple MirrorDNA client for local operations.
 *
 * Features:
 * - Load and validate vault configurations
 * - Compute deterministic state hashes for directories
 * - Validate timeline JSON/YAML files
 *
 * All operations are local-only using Node.js built-in modules.
 */
class MirrorDNAClient {
    constructor() {
        this.lastVaultConfig = null;
        this.lastStateHash = null;
    }

    /**
     * Load and validate a vault configuration file.
     *
     * @param {string} filePath - Path to vault config file (JSON or YAML)
     * @returns {Object} Vault configuration object
     * @throws {Error} If file doesn't exist or config is invalid
     *
     * @example
     * const client = new MirrorDNAClient();
     * const config = client.loadVaultConfig('vault.json');
     * console.log(config.vault_id);
     */
    loadVaultConfig(filePath) {
        if (!fs.existsSync(filePath)) {
            throw new Error(`Vault config not found: ${filePath}`);
        }

        // Read file
        const content = fs.readFileSync(filePath, 'utf8');
        let data;

        // Parse based on extension
        const ext = path.extname(filePath).toLowerCase();
        if (ext === '.yaml' || ext === '.yml') {
            // For YAML, try to use js-yaml if available, otherwise throw
            try {
                const yaml = require('js-yaml');
                data = yaml.load(content);
            } catch (e) {
                throw new Error('js-yaml required for YAML files. Install with: npm install js-yaml');
            }
        } else {
            data = JSON.parse(content);
        }

        // Basic validation
        const requiredFields = ['vault_id', 'name', 'path', 'created_at'];
        const missing = requiredFields.filter(field => !(field in data));

        if (missing.length > 0) {
            throw new Error(`Vault config missing required fields: ${missing.join(', ')}`);
        }

        this.lastVaultConfig = data;
        return data;
    }

    /**
     * Compute deterministic SHA-256 hash of directory contents.
     *
     * Creates a hash based on file paths and contents in alphabetical order.
     * Useful for detecting changes in vault state.
     *
     * @param {string} directory - Path to directory to hash
     * @param {string[]} ignorePatterns - Optional patterns to ignore
     * @returns {string} Hexadecimal SHA-256 hash string
     * @throws {Error} If directory doesn't exist
     *
     * @example
     * const client = new MirrorDNAClient();
     * const hash1 = client.computeStateHash('./my_vault');
     * // ... make changes ...
     * const hash2 = client.computeStateHash('./my_vault');
     * if (hash1 !== hash2) {
     *     console.log('Vault state changed!');
     * }
     */
    computeStateHash(directory, ignorePatterns = null) {
        if (!fs.existsSync(directory)) {
            throw new Error(`Directory not found: ${directory}`);
        }

        const stats = fs.statSync(directory);
        if (!stats.isDirectory()) {
            throw new Error(`Path is not a directory: ${directory}`);
        }

        // Default ignore patterns
        if (!ignorePatterns) {
            ignorePatterns = ['.git', 'node_modules', '__pycache__', '.DS_Store', '.pyc'];
        }

        // Collect all file hashes
        const fileHashes = [];
        this._walkDirectory(directory, directory, fileHashes, ignorePatterns);

        // Sort for determinism
        fileHashes.sort();

        // Compute final hash
        const combined = fileHashes.join('\n');
        const hash = crypto.createHash('sha256').update(combined).digest('hex');

        this.lastStateHash = hash;
        return hash;
    }

    /**
     * Recursively walk directory and collect file hashes.
     * @private
     */
    _walkDirectory(baseDir, currentDir, fileHashes, ignorePatterns) {
        const entries = fs.readdirSync(currentDir);

        for (const entry of entries) {
            // Check if should ignore
            if (this._shouldIgnore(entry, ignorePatterns)) {
                continue;
            }

            const fullPath = path.join(currentDir, entry);
            const stats = fs.statSync(fullPath);

            if (stats.isDirectory()) {
                this._walkDirectory(baseDir, fullPath, fileHashes, ignorePatterns);
            } else if (stats.isFile()) {
                try {
                    // Compute file hash
                    const content = fs.readFileSync(fullPath);
                    const fileHash = crypto.createHash('sha256').update(content).digest('hex');

                    // Get relative path
                    const relativePath = path.relative(baseDir, fullPath);

                    // Add to list
                    fileHashes.push(`${relativePath}:${fileHash}`);
                } catch (e) {
                    // Skip files that can't be read
                    continue;
                }
            }
        }
    }

    /**
     * Check if filename/dirname matches any ignore pattern.
     * @private
     */
    _shouldIgnore(name, patterns) {
        for (const pattern of patterns) {
            if (pattern.startsWith('*')) {
                if (name.endsWith(pattern.substring(1))) {
                    return true;
                }
            } else if (name.includes(pattern)) {
                return true;
            }
        }
        return false;
    }

    /**
     * Validate timeline file structure and return summary.
     *
     * Checks that timeline file contains required fields and valid event structure.
     *
     * @param {string} filePath - Path to timeline file (JSON or YAML)
     * @returns {Object} Validation results with timeline summary
     *
     * @example
     * const client = new MirrorDNAClient();
     * const result = client.validateTimeline('timeline.json');
     * if (result.valid) {
     *     console.log(`Timeline valid with ${result.event_count} events`);
     * }
     */
    validateTimeline(filePath) {
        const errors = [];

        if (!fs.existsSync(filePath)) {
            return {
                valid: false,
                errors: [`File not found: ${filePath}`]
            };
        }

        // Load timeline file
        let data;
        try {
            const content = fs.readFileSync(filePath, 'utf8');
            const ext = path.extname(filePath).toLowerCase();

            if (ext === '.yaml' || ext === '.yml') {
                try {
                    const yaml = require('js-yaml');
                    data = yaml.load(content);
                } catch (e) {
                    return {
                        valid: false,
                        errors: ['js-yaml required for YAML files']
                    };
                }
            } else {
                data = JSON.parse(content);
            }
        } catch (e) {
            return {
                valid: false,
                errors: [`Failed to parse file: ${e.message}`]
            };
        }

        // Extract events
        let events, timelineId;
        if (Array.isArray(data)) {
            events = data;
            timelineId = 'unknown';
        } else if (typeof data === 'object') {
            events = data.events || [];
            timelineId = data.timeline_id || 'unknown';
        } else {
            return {
                valid: false,
                errors: ["Timeline file must be an array or object with 'events' key"]
            };
        }

        // Validate event structure
        const requiredEventFields = ['id', 'timestamp', 'event_type', 'actor'];

        events.forEach((event, i) => {
            if (typeof event !== 'object' || event === null) {
                errors.push(`Event ${i} is not an object`);
                return;
            }

            const missing = requiredEventFields.filter(field => !(field in event));
            if (missing.length > 0) {
                errors.push(`Event ${i} missing fields: ${missing.join(', ')}`);
            }
        });

        // Build summary
        const result = {
            valid: errors.length === 0,
            event_count: events.length,
            timeline_id: timelineId,
            errors: errors
        };

        if (events.length > 0) {
            result.first_event = events[0].timestamp || 'unknown';
            result.last_event = events[events.length - 1].timestamp || 'unknown';
        }

        return result;
    }

    /**
     * Compute deterministic SHA-256 checksum for object data.
     *
     * Uses canonical JSON serialization (sorted keys) for determinism.
     *
     * @param {Object} data - Object to hash
     * @returns {string} Hexadecimal SHA-256 hash string
     *
     * @example
     * const client = new MirrorDNAClient();
     * const checksum = client.computeDataChecksum({id: 'test', value: 42});
     */
    computeDataChecksum(data) {
        // Create canonical JSON (sorted keys)
        const canonicalJson = JSON.stringify(data, Object.keys(data).sort());
        return crypto.createHash('sha256').update(canonicalJson).digest('hex');
    }

    /**
     * Get continuity status summary for vault and timeline.
     *
     * Convenience method that combines vault config, state hash, and timeline validation.
     *
     * @param {string} vaultPath - Optional path to vault config
     * @param {string} timelinePath - Optional path to timeline file
     * @returns {Object} Continuity status information
     *
     * @example
     * const client = new MirrorDNAClient();
     * const status = client.getContinuityStatus({
     *     vaultPath: 'vault.json',
     *     timelinePath: 'timeline.json'
     * });
     * console.log(`Vault: ${status.vault_id}, Events: ${status.event_count}`);
     */
    getContinuityStatus({ vaultPath = null, timelinePath = null } = {}) {
        const status = {
            timestamp: new Date().toISOString(),
            vault_loaded: false,
            timeline_valid: false
        };

        if (vaultPath) {
            try {
                const vault = this.loadVaultConfig(vaultPath);
                status.vault_loaded = true;
                status.vault_id = vault.vault_id;
                status.vault_name = vault.name;
            } catch (e) {
                status.vault_error = e.message;
            }
        }

        if (timelinePath) {
            const result = this.validateTimeline(timelinePath);
            status.timeline_valid = result.valid;
            status.event_count = result.event_count || 0;
            status.timeline_errors = result.errors || [];
        }

        // Compute state hash if we have a vault
        if (vaultPath && status.vault_loaded && this.lastVaultConfig) {
            try {
                const vaultDir = this.lastVaultConfig.path;
                if (fs.existsSync(vaultDir)) {
                    status.state_hash = this.computeStateHash(vaultDir);
                }
            } catch (e) {
                status.state_hash_error = e.message;
            }
        }

        return status;
    }
}

// Convenience functions for quick operations

/**
 * Quick utility to hash a directory.
 *
 * @param {string} directory - Path to directory
 * @returns {string} SHA-256 hash string
 */
function quickHashDirectory(directory) {
    const client = new MirrorDNAClient();
    return client.computeStateHash(directory);
}

/**
 * Quick utility to check if timeline is valid.
 *
 * @param {string} filePath - Path to timeline file
 * @returns {boolean} True if valid, false otherwise
 */
function quickValidateTimeline(filePath) {
    const client = new MirrorDNAClient();
    const result = client.validateTimeline(filePath);
    return result.valid;
}

// Exports
module.exports = {
    MirrorDNAClient,
    quickHashDirectory,
    quickValidateTimeline
};

File: sdk/python/mirrordna_client.py
"""
MirrorDNA Developer SDK - Python Client

Simple, local-only client for MirrorDNA protocol operations.
Provides easy-to-use interface for vault config loading, state hashing, and timeline validation.

This SDK is designed for:
- Local development and testing
- Understanding MirrorDNA concepts
- Building simple integrations

For production use, see the full protocol implementation in src/mirrordna/
"""

import os
import hashlib
import json
from pathlib import Path
from typing import Dict, Any, Optional, Union


class MirrorDNAClient:
    """
    Simple MirrorDNA client for local operations.

    Features:
    - Load and validate vault configurations
    - Compute deterministic state hashes for directories
    - Validate timeline JSON/YAML files

    All operations are local-only with no external dependencies beyond pyyaml.
    """

    def __init__(self):
        """Initialize MirrorDNA client."""
        self.last_vault_config = None
        self.last_state_hash = None

    def load_vault_config(self, path: str) -> Dict[str, Any]:
        """
        Load and validate a vault configuration file.

        Args:
            path: Path to vault config file (JSON or YAML)

        Returns:
            Dictionary containing vault configuration

        Raises:
            FileNotFoundError: If file doesn't exist
            ValueError: If config is invalid

        Example:
            >>> client = MirrorDNAClient()
            >>> config = client.load_vault_config("vault.yaml")
            >>> print(config['vault_id'])
        """
        path = Path(path)

        if not path.exists():
            raise FileNotFoundError(f"Vault config not found: {path}")

        # Load file based on extension
        with open(path, 'r') as f:
            if path.suffix in ['.yaml', '.yml']:
                try:
                    import yaml
                    data = yaml.safe_load(f)
                except ImportError:
                    raise ImportError(
                        "PyYAML required for YAML files. Install with: pip install pyyaml"
                    )
            else:
                data = json.load(f)

        # Basic validation
        required_fields = ['vault_id', 'name', 'path', 'created_at']
        missing = [f for f in required_fields if f not in data]

        if missing:
            raise ValueError(f"Vault config missing required fields: {missing}")

        self.last_vault_config = data
        return data

    def compute_state_hash(self, directory: str, ignore_patterns: Optional[list] = None) -> str:
        """
        Compute deterministic SHA-256 hash of directory contents.

        Creates a hash based on file paths and contents in alphabetical order.
        Useful for detecting changes in vault state.

        Args:
            directory: Path to directory to hash
            ignore_patterns: Optional list of patterns to ignore (e.g., ['.git', '__pycache__'])

        Returns:
            Hexadecimal SHA-256 hash string

        Raises:
            FileNotFoundError: If directory doesn't exist

        Example:
            >>> client = MirrorDNAClient()
            >>> hash1 = client.compute_state_hash("./my_vault")
            >>> # Make changes to files...
            >>> hash2 = client.compute_state_hash("./my_vault")
            >>> if hash1 != hash2:
            ...     print("Vault state has changed!")
        """
        directory = Path(directory)

        if not directory.exists():
            raise FileNotFoundError(f"Directory not found: {directory}")

        if not directory.is_dir():
            raise ValueError(f"Path is not a directory: {directory}")

        # Default ignore patterns
        if ignore_patterns is None:
            ignore_patterns = ['.git', '__pycache__', '.DS_Store', '*.pyc']

        # Collect all files in deterministic order
        file_hashes = []

        for root, dirs, files in os.walk(directory):
            # Filter out ignored directories
            dirs[:] = [d for d in dirs if not self._should_ignore(d, ignore_patterns)]

            # Process files in alphabetical order
            for filename in sorted(files):
                if self._should_ignore(filename, ignore_patterns):
                    continue

                filepath = Path(root) / filename
                rel_path = filepath.relative_to(directory)

                # Hash file content
                try:
                    with open(filepath, 'rb') as f:
                        file_hash = hashlib.sha256(f.read()).hexdigest()

                    # Combine relative path and content hash
                    file_hashes.append(f"{rel_path}:{file_hash}")
                except (IOError, OSError):
                    # Skip files that can't be read
                    continue

        # Compute final hash from all file hashes
        combined = '\n'.join(sorted(file_hashes))
        final_hash = hashlib.sha256(combined.encode('utf-8')).hexdigest()

        self.last_state_hash = final_hash
        return final_hash

    def _should_ignore(self, name: str, patterns: list) -> bool:
        """Check if filename/dirname matches any ignore pattern."""
        for pattern in patterns:
            if pattern.startswith('*'):
                if name.endswith(pattern[1:]):
                    return True
            elif pattern in name:
                return True
        return False

    def validate_timeline(self, path: str) -> Dict[str, Any]:
        """
        Validate timeline file structure and return summary.

        Checks that timeline file contains required fields and valid event structure.
        Does not perform deep schema validation.

        Args:
            path: Path to timeline file (JSON or YAML)

        Returns:
            Dictionary with validation results and timeline summary:
            {
                'valid': bool,
                'event_count': int,
                'timeline_id': str,
                'errors': list,
                'first_event': str,
                'last_event': str
            }

        Example:
            >>> client = MirrorDNAClient()
            >>> result = client.validate_timeline("timeline.json")
            >>> if result['valid']:
            ...     print(f"Timeline valid with {result['event_count']} events")
            >>> else:
            ...     print(f"Errors: {result['errors']}")
        """
        path = Path(path)
        errors = []

        if not path.exists():
            return {
                'valid': False,
                'errors': [f"File not found: {path}"]
            }

        # Load timeline file
        try:
            with open(path, 'r') as f:
                if path.suffix in ['.yaml', '.yml']:
                    try:
                        import yaml
                        data = yaml.safe_load(f)
                    except ImportError:
                        return {
                            'valid': False,
                            'errors': ["PyYAML required for YAML files"]
                        }
                else:
                    data = json.load(f)
        except Exception as e:
            return {
                'valid': False,
                'errors': [f"Failed to parse file: {e}"]
            }

        # Extract events
        if isinstance(data, list):
            events = data
            timeline_id = "unknown"
        elif isinstance(data, dict):
            events = data.get('events', [])
            timeline_id = data.get('timeline_id', 'unknown')
        else:
            return {
                'valid': False,
                'errors': ["Timeline file must be a list or dict with 'events' key"]
            }

        # Validate event structure
        required_event_fields = ['id', 'timestamp', 'event_type', 'actor']

        for i, event in enumerate(events):
            if not isinstance(event, dict):
                errors.append(f"Event {i} is not a dictionary")
                continue

            missing = [f for f in required_event_fields if f not in event]
            if missing:
                errors.append(f"Event {i} missing fields: {missing}")

        # Build summary
        result = {
            'valid': len(errors) == 0,
            'event_count': len(events),
            'timeline_id': timeline_id,
            'errors': errors
        }

        if events:
            result['first_event'] = events[0].get('timestamp', 'unknown')
            result['last_event'] = events[-1].get('timestamp', 'unknown')

        return result

    def compute_data_checksum(self, data: Dict[str, Any]) -> str:
        """
        Compute deterministic SHA-256 checksum for dictionary data.

        Uses canonical JSON serialization (sorted keys) for determinism.

        Args:
            data: Dictionary to hash

        Returns:
            Hexadecimal SHA-256 hash string

        Example:
            >>> client = MirrorDNAClient()
            >>> checksum = client.compute_data_checksum({'id': 'test', 'value': 42})
        """
        canonical_json = json.dumps(data, sort_keys=True, separators=(',', ':'))
        return hashlib.sha256(canonical_json.encode('utf-8')).hexdigest()

    def get_continuity_status(
        self,
        vault_path: Optional[str] = None,
        timeline_path: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Get continuity status summary for vault and timeline.

        Convenience method that combines vault config, state hash, and timeline validation.

        Args:
            vault_path: Optional path to vault config
            timeline_path: Optional path to timeline file

        Returns:
            Dictionary with continuity status information

        Example:
            >>> client = MirrorDNAClient()
            >>> status = client.get_continuity_status(
            ...     vault_path="vault.yaml",
            ...     timeline_path="timeline.json"
            ... )
            >>> print(f"Vault: {status['vault_id']}, Events: {status['event_count']}")
        """
        status = {
            'timestamp': self._get_timestamp(),
            'vault_loaded': False,
            'timeline_valid': False
        }

        if vault_path:
            try:
                vault = self.load_vault_config(vault_path)
                status['vault_loaded'] = True
                status['vault_id'] = vault.get('vault_id')
                status['vault_name'] = vault.get('name')
            except Exception as e:
                status['vault_error'] = str(e)

        if timeline_path:
            result = self.validate_timeline(timeline_path)
            status['timeline_valid'] = result['valid']
            status['event_count'] = result.get('event_count', 0)
            status['timeline_errors'] = result.get('errors', [])

        # Compute state hash if we have a vault path
        if vault_path and status['vault_loaded']:
            try:
                vault_dir = Path(self.last_vault_config.get('path', '.'))
                if vault_dir.exists():
                    status['state_hash'] = self.compute_state_hash(str(vault_dir))
            except Exception as e:
                status['state_hash_error'] = str(e)

        return status

    def _get_timestamp(self) -> str:
        """Get current UTC timestamp in ISO format."""
        from datetime import datetime
        return datetime.utcnow().isoformat() + 'Z'


# Convenience functions for quick operations

def quick_hash_directory(directory: str) -> str:
    """
    Quick utility to hash a directory.

    Args:
        directory: Path to directory

    Returns:
        SHA-256 hash string
    """
    client = MirrorDNAClient()
    return client.compute_state_hash(directory)


def quick_validate_timeline(path: str) -> bool:
    """
    Quick utility to check if timeline is valid.

    Args:
        path: Path to timeline file

    Returns:
        True if valid, False otherwise
    """
    client = MirrorDNAClient()
    result = client.validate_timeline(path)
    return result['valid']
